Services/DevicePropertiesRequest.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services
+{
+    public interface IDevicePropertiesRequest
+    {
+        Task RegisterDevicePropertiesUpdateAsync(
+            string deviceId,
+            ISmartDictionary deviceProperties);
+
+        Task OnPropertyUpdateRequestedCallback(
+            TwinCollection desiredProperties,
+            object userContext);
+    }
+
+    public class DevicePropertiesRequest : IDevicePropertiesRequest
+    {
+        private readonly Azure.Devices.Client.DeviceClient client;
+        private readonly ILogger log;
+        private string deviceId;
+        private ISmartDictionary deviceProperties;
+
+        public DevicePropertiesRequest(Azure.Devices.Client.DeviceClient client, ILogger logger)
+        {
+            this.client = client;
+            this.log = logger;
+            this.deviceId = string.Empty;
+        }
+
+        public async Task RegisterDevicePropertiesUpdateAsync(string deviceId, ISmartDictionary deviceProperties)
+        {
+            if (this.deviceId != string.Empty)
+            {
+                this.log.Error("Application error, each device must have a separate instance", () => { });
+                throw new Exception("Application error, each device must have a separate instance of " + this.GetType().FullName);
+            }
```

Comment:
curious, how would the setup get into this state ? Is this check necessary ?

Reply:
It's a preventative measure in case the consumer of this class were to call registration more than once. Each device should have a separate instance of `DeviceClient` which has `DevicePropertiesRequest` that does the registration with the azure device client only once. If the deviceId were already set then this class could accidentally update the state on the callback for the wrong device on the hub.